pkg/manager/config/config.go

```diff
@@ -15,44 +15,80 @@
 package config
 
 import (
-	"context"
-	"encoding/json"
-	"path"
+	"os"
+	"time"
 
+	"github.com/BurntSushi/toml"
+	"github.com/fsnotify/fsnotify"
 	"github.com/pingcap/TiProxy/lib/config"
-	"github.com/pingcap/TiProxy/lib/util/errors"
 	"go.uber.org/zap"
 )
 
-func (e *ConfigManager) SetConfig(ctx context.Context, val *config.Config) error {
-	c, err := json.Marshal(val)
+func (e *ConfigManager) reloadConfigFile(file string) error {
+	proxyConfigData, err := os.ReadFile(file)
 	if err != nil {
-		return errors.WithStack(err)
+		return err
 	}
-	return e.set(ctx, pathPrefixConfig, "all", c)
+
+	return e.SetTOMLConfig(proxyConfigData)
 }
 
-func (e *ConfigManager) GetConfig(ctx context.Context, val *config.Config) error {
-	c, err := e.get(ctx, pathPrefixConfig, "all")
-	if err != nil {
+func (e *ConfigManager) handleFSEvent(ev fsnotify.Event) {
+	switch {
+	case ev.Has(fsnotify.Create), ev.Has(fsnotify.Write), ev.Has(fsnotify.Remove), ev.Has(fsnotify.Rename):
+		// in case of remove/rename the file
+		if ev.Has(fsnotify.Remove) || ev.Has(fsnotify.Rename) {
+			// rewatching may be too fast, sleep for a while
+			time.Sleep(50 * time.Millisecond)
+
+			nerr := e.wch.Add(ev.Name)
+			if nerr != nil {
+				e.logger.Error("failed to re-watch file", zap.String("file", ev.Name), zap.Error(nerr))
+			}
+		}
+
+		// try to reload it
+		e.logger.Info("config file reloaded", zap.Error(e.reloadConfigFile(ev.Name)), zap.Any("cfg", e.GetConfig()))
+	}
+}
+
+func (e *ConfigManager) SetTOMLConfig(data []byte) error {
```

Comment:
It is not `SetConfig(*config.Config)`, because, we can tell if user defined`proxy.max-connections = 0` or empty. The former should set the value to zero, while the latter should keep things unchanged.

Reply:
Because the input is TOML string, which is not a struct. We known you wrote `= 0` or just empty.

Reply:
You see, the thing is:

I only need to modify the current config struct `max-conns = 30` if you wrote `max-conns = 0` but empty `""`. But config struct `max-conns = 0` always means no limits.

Reply:
There are locks, `e.sts.Lock()`.

Reply:
goroutine 1:
cfg := cfgMgr.GetConfig()
// read cfg     <- unsafe

goroutine 2:
SetTOMLConfig(newCfg)

Reply:
Now it is cloned in `SetTOMLConfig`, then the one modified by `SetTOMLConfig` and the one returned by `GetConfig` are different.